### DISCUSSION TITLE
How does `task_local!` work?

### DISCUSSION THREAD

**Discussion by jjant** [tokio-rs/tokio#6111] [Q&A]
If I understand correctly, task locals are like thread locals, but they handle:

- The same thread possibly interleaving many tasks (a `ThreadLocal` would be overwritten)
- The same task switching threads (the `ThreadLocal` being unset for the task)

Can anyone explain the technical details of Tokio's implementation of `task_local!`?

**Comment by Darksonn** [comment]
Most of the time, the value is stored inside the task local scope future. However, when you poll the scope future, it will temporarily move the value to a thread-local. When the poll on the scope future ends, the value is moved out of the thread-local.

This means that whenever you are inside a call to poll on the task-local scope future, the value will be in the right thread-local. This allows you to access it. You can't access the value outside of the scope future, but that's okay.
